crossbeam/property_signatures/property_signatures.py

```diff
@@ -0,0 +1,474 @@
+    fixed_length: bool = True) -> List[Optional[bool]]:
+  """Returns a property signature for an object in context of an I/O example."""
+  return _basic_signature(x, fixed_length) + _compare(x, output, fixed_length)
+
+
+def _reduce_across_examples(
+    signatures: List[List[Optional[bool]]]
+) -> List[Tuple[float, bool, bool, float]]:
+  """Reduce across examples (frac applicable, all True?, all False?, frac True)."""
+  # All signatures should have the same length across examples.
+  num_examples = len(signatures)
+  assert num_examples
+  signature_len = len(signatures[0])
+  assert all(len(sig) == signature_len for sig in signatures)
+
+  result = []
+  for bools in zip(*signatures):
+    bools_not_none = [x for x in bools if x is not None]
+    num_not_none = len(bools_not_none)
+    frac_true = sum(bools_not_none) / num_not_none if num_not_none else 0.5
+    result.append((len(bools_not_none) / num_examples,
+                   frac_true == 1,
+                   frac_true == 0,
+                   frac_true))
+  return result
+
+
+def property_signature_io_examples(
+    input_values: List[value_module.Value],
+    output_value: value_module.Value,
+    fixed_length: bool = True) -> List[Tuple[float, bool, bool, float]]:
+  """Returns a property signature for a set of I/O examples."""
+  num_examples = output_value.num_examples
+  assert all(i_value.num_examples == num_examples for i_value in input_values)
+  return _reduce_across_examples(
+      [_property_signature_single_example(  # pylint: disable=g-complex-comprehension
+          [i_value[example_index] for i_value in input_values],
+          output_value[example_index], fixed_length=fixed_length)
+       for example_index in range(num_examples)])
+
+IO_EXAMPLES_SIGNATURE_LENGTH = len(property_signature_io_examples(
+    [value_module.InputVariable([1, 2], 'in1')],
+    value_module.OutputValue([-1, -2]),
+    fixed_length=True))
+
+
+def _property_signature_concrete_value(
+    value: value_module.Value,
+    output_value: value_module.Value,
+    fixed_length: bool = True) -> List[Tuple[float, bool, bool, float]]:
+  """Returns a property signature for a value w.r.t. a set of I/O examples."""
+  assert value.num_free_variables == 0
+  return _reduce_across_examples(
+      [_property_signature_single_object(
+          value[i], output_value[i], fixed_length=fixed_length)
+       for i in range(output_value.num_examples)])
+
+
+def _run_lambda(value: value_module.Value) -> List[List[Tuple[List[Any], Any]]]:
+  """Runs a lambda on canonical values."""
+  arity = value.num_free_variables
+  assert arity > 0
+  num_examples = value.num_examples
+  io_pairs_per_example = [[] for _ in range(num_examples)]
+  inputs_to_try = sum(VALUES_TO_TRY.values(), [])
+  for inputs_list in itertools.product(inputs_to_try, repeat=arity):
+    for lambda_fn, io_pairs in zip(value.values, io_pairs_per_example):
+      try:
+        result = lambda_fn(*inputs_list)
+        if result is not None:
```

Comment:
minor: maybe we can also record down the inapplicable inputs_list (or the types) and serve as auxiliary features.

Reply:
the types end up participating in the property signature (through _type_property) already. do you mean we should collect them explicitly as an additional return value?

Reply:
I was thinking probably the way it fails (either throwing error or returning none) might indicate something, but anyway this is not in need at the moment.